README.md
# [update_app](https://pub.dev/packages/update_app)

Update the app, pass in the updated address, and update the app. 更新app, 传入更新的地址, 进行app更新

The Android part is implemented using DownloadManager. If the file has been downloaded, install it directly.
Android部分使用DownloadManager实现.若文件已下载, 则直接安装.

The Ios section does not support external updates, skip the App Store. Ios部分不支持外部更新, 跳转App Store.

## Experience demo(体验Demo)

[download apk(下载APK)](apks/app-release.apk)

 <img src="https://raw.githubusercontent.com/mofada/flutter_update_app/master/images/screen.png" width = "320" height = "580" alt="screen" align=center />

## How to use(如何使用)

```dart
import 'package:update_app/update_app.dart';

/// Download the app and update
///
/// appleId: This is the app id of the app store, used to jump to the app store
/// url: This is the URL to download the apk file
/// title: Android download the title displayed in the notification bar, generally use the file name, or like this `update version 1.6`
/// description: Android download the description displayed in the notification bar, similar to the subtitle
///
/// @return `-1` -> `download failed`, `0` -> `There is a current apk locally, and the download is successful`, other (int) -> `downloadId`
var downloadId = await
UpdateApp.updateApp(url: "apkPath", appleId:"375380948", title:"通知标题", description:"通知描述");

/// Query apk download progress
///
/// downloadId: [updateApp] 返回的downloadId
///
/// @return [DownloadProcess]
/// current: Currently downloaded length
/// count: Total length
/// status: [ProcessState]
///         STATUS_PENDING: [ProcessState.STATUS_PENDING](https://developer.android.com/reference/android/app/DownloadManager#STATUS_PENDING)
///         STATUS_RUNNING: [ProcessState.STATUS_RUNNING](https://developer.android.com/reference/android/app/DownloadManager#STATUS_RUNNING)
///         STATUS_PAUSED: [ProcessState.STATUS_PAUSED](https://developer.android.com/reference/android/app/DownloadManager#STATUS_PAUSED)
///         STATUS_SUCCESSFUL: [ProcessState.STATUS_SUCCESSFUL](https://developer.android.com/reference/android/app/DownloadManager#STATUS_SUCCESSFUL)
///         STATUS_FAILED: [ProcessState.STATUS_FAILED](https://developer.android.com/reference/android/app/DownloadManager#STATUS_FAILED)
var process = await
UpdateApp.downloadProcess(downloadId: downloadId);
```

## Document

### UpdateApp.updateApp

#### Params

appleId - This is the app id of the app store, used to jump to the app store  
url - This is the URL to download the apk file  
title - Android download the title displayed in the notification bar, generally use the file name description - Android
download the description displayed in the notification bar, similar to the subtitle

#### Return

`-1` -> `download failed`   
`0` -> `There is a current apk locally, and the download is successful`   
other (int) -> `downloadId`

### UpdateApp.downloadProcess

#### Params

downloadId - [updateApp] 返回的downloadId

#### Return
[DownloadProcess]   
current: Currently downloaded length  
count: Total length  
status: [ProcessState]  
STATUS_PENDING: [ProcessState.STATUS_PENDING](https://developer.android.com/reference/android/app/DownloadManager#STATUS_PENDING)  
STATUS_RUNNING: [ProcessState.STATUS_RUNNING](https://developer.android.com/reference/android/app/DownloadManager#STATUS_RUNNING)  
STATUS_PAUSED: [ProcessState.STATUS_PAUSED](https://developer.android.com/reference/android/app/DownloadManager#STATUS_PAUSED)  
STATUS_SUCCESSFUL: [ProcessState.STATUS_SUCCESSFUL](https://developer.android.com/reference/android/app/DownloadManager#STATUS_SUCCESSFUL)  
STATUS_FAILED: [ProcessState.STATUS_FAILED](https://developer.android.com/reference/android/app/DownloadManager#STATUS_FAILED)  

# Features(特点)

## Android part(Android部分)

1. Download using [DownloadManager](https://developer.android.com/reference/android/app/DownloadManager)  
   使用[DownloadManager](https://developer.android.com/reference/android/app/DownloadManager)进行下载
2. Monitor download completion broadcast  
   监听下载完成广播
3. Adapt [Android 6.0 Runtime Permissions](https://developer.android.com/training/permissions/requesting?hl=en) Files
   are stored in the app directory, no need to apply for runtime permissions  
   适配[Android 6.0运行时权限](https://developer.android.com/training/permissions/requesting?hl=zh-cn) 文件存储在app目录, 不需要申请运行时权限
4. Adapt [Android 7.0FileProvider](https://developer.android.com/reference/android/support/v4/content/FileProvider)  
   适配[Android 7.0FileProvider](https://developer.android.com/reference/android/support/v4/content/FileProvider)
5. Adapt Android 8.0 installation permissions  
   适配Android 8.0安装权限
6.
Adapt [Android 9.0 Network Security Configuration] (https://developer.android.com/training/articles/security-config)  
适配[Android 9.0网络安全配置](https://developer.android.com/training/articles/security-config)

## IOS section(IOS部分)

1. Jump to the app store 跳转应用商店

CHANGELOG.md
## 0.0.1
1. 使用[DownloadManager](https://developer.android.com/reference/android/app/DownloadManager)进行下载
2. 监听下载完成广播
3. 适配[Android 6.0运行时权限]() 文件存储在app目录, 不需要申请运行时权限
4. 适配[Android 7.0FileProvider](https://developer.android.com/reference/android/support/v4/content/FileProvider)
5. 适配[Android 8.0安装权限]()
6. 适配[Android 9.0网络安全通信]()

## 0.0.2
未修改

## 0.0.3
1. 完成IOS部分, 跳转Apple Store

## 0.0.4
1. 修复bug

## 0.0.5
1.修改版本问题

## 0.1.0
1.Modify the prompt and version number according to pub.dev

## 0.1.1
1.update README.MD

## 0.1.2
1.update README.MD

## 0.1.3
1.fix issues[#3](https://github.com/mofada/flutter_update_app/issues/3)

## 0.1.4
1. fix issues[#5](https://github.com/mofada/flutter_update_app/issues/5)
2.change gradle version

## 0.1.5
1.update yaml description

## 0.1.6
1. Merge pull request #9

## 0.1.7
1. Fix #1
2. Add download progress monitoring


example/lib/main.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:update_app/bean/download_process.dart';

import 'package:update_app/update_app.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  //定时更新进度
  Timer? timer;

  //下载进度
  double downloadProcess = 0;

  //下载状态
  String downloadStatus = "";

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext? context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Update app'),
        ),
        body: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Container(
                padding: EdgeInsets.all(16),
                width: 200,
                height: 200,
                child: CircularProgressIndicator(
                  value: downloadProcess,
                  strokeWidth: 10,
                ),
              ),
              Text("Download status: $downloadStatus"),
              Text("Please replace the download url with your own address"),
            ],
          ),
        ),
        bottomNavigationBar: Container(
            padding: EdgeInsets.all(20),
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(
                  'click right bottom download button download!',
                  style: TextStyle(color: Theme.of(context!).primaryColor),
                ),
                Text('Adapt to Android 6.0'),
                Text('Adapt to Android 7.0'),
                Text('Adapt to Android 8.0'),
                Text('Adapt to Android 9.0'),
                Text('Adapt to Android 10.0'),
                Text('Adapt to Android 10.0'),
                Text('Add download progress monitoring'),
              ],
            )),
        floatingActionButton: FloatingActionButton(
          onPressed: download,
          child: Icon(Icons.file_download),
        ),
        floatingActionButtonLocation: FloatingActionButtonLocation.endDocked,
      ),
    );
  }

  @override
  void dispose() {
    timer?.cancel();
    super.dispose();
  }

  void download() async {
    var downloadId = await UpdateApp.updateApp(
        url:
            "https://raw.githubusercontent.com/mofada/flutter_update_app/master/apks/app-release.apk",
        appleId: "375380948");

    //本地已有一样的apk, 下载成功
    if (downloadId == 0) {
      setState(() {
        downloadProcess = 1;
        downloadStatus = ProcessState.STATUS_SUCCESSFUL.toString();
      });
      return;
    }

    //出现了错误, 下载失败
    if (downloadId == -1) {
      setState(() {
        downloadProcess = 1;
        downloadStatus = ProcessState.STATUS_FAILED.toString();
      });
      return;
    }

    //正在下载文件
    timer = Timer.periodic(Duration(milliseconds: 100), (timer) async {
      var process = await UpdateApp.downloadProcess(downloadId: downloadId);
      //更新界面状态
      setState(() {
        downloadProcess = process.current / process.count;
        downloadStatus = process.status.toString();
      });

      if (process.status == ProcessState.STATUS_SUCCESSFUL ||
          process.status == ProcessState.STATUS_FAILED) {
        //如果已经下载成功, 取消计时
        timer.cancel();
      }
    });
  }
}
